File: src/solver.py
# Copyright 2024 Bang-Shien Chen.
# All rights reserved. See LICENSE for the license information.

import numpy as np
import scipy as sp


class MAP:
    def __init__(
        self, max_iteration, tolerance, relative_tolerance=1e-6, verbose=False
    ):
        """
        Method of Alternating Projections.

        Arguments

        - `max_iteration`      - Maximum number of iterations.
        - `tolerance`          - Absolute tolerance for the error.
        - `relative_tolerance` - Relative tolerance for early stop.
        - `verbose`            - Verbosity.
        """
        self.max_iter = max_iteration
        self.tol = tolerance
        self.rel_tol = relative_tolerance
        self.verbose = verbose

    def solve(self, mat_A, mat_B, vec_c, initial=None):
        """
        Solve an Absolute Value Equation

            $$ Ax + B|x| = c $$

        using the Method of Alternating Projections.

        Arguments

        - `mat_A`   - Matrix $A$ of size (m, n).
        - `mat_B`   - Matrix $B$ of size (m, n).
        - `vec_c`   - Vector $c$ of size (m,).
        - `initial` - Initial guess vector of size (2n,).

        Returns

        - `vec_x` - Solution vector of size (n,).
        """
        m, n = mat_A.shape
        mat_T = np.hstack([mat_A + mat_B, -mat_A + mat_B])
        cbar = np.sqrt(2) * vec_c

        # check full rank case by Lemma 2.8
        case_full_rank = False
        if m == n and np.allclose(mat_B, -np.eye(3)):
            case_full_rank = True

        # pre-compute cholesky factor for efficiency (page. 29)
        chol_factor, chol_bool = sp.linalg.cho_factor(
            2 * (mat_A @ mat_A.T + mat_B @ mat_B.T)
        )

        # initialization
        if initial is None:
            vec_w = mat_T.T @ sp.linalg.cho_solve((chol_factor, chol_bool), cbar)
        else:
            assert (
                len(initial) == 2 * n
            ), "dimension of initial guess must be {}".format(2 * n)
            vec_w = initial

        prev_cost = 1e10
        for i in range(self.max_iter):
            # projection onto C2 (Proposition 2.2)
            vec_u = vec_w[:n]  # first-half of `vec_w`
            vec_v = vec_w[n:]  # second-half of `vec_w`
            mask = (vec_u < 0) & (vec_v < 0)  # both negative
            vec_u = np.where(mask, 0, np.where(vec_u >= vec_v, vec_u, 0))
            vec_v = np.where(mask, 0, np.where(vec_v >= vec_u, vec_v, 0))
            pc2_w = np.hstack([vec_u, vec_v])

            # projection onto C1 (Proposition 2.1)
            # solve `vec_z` by Equation (4.1)
            vec_z = sp.linalg.cho_solve((chol_factor, chol_bool), mat_T @ pc2_w - cbar)

            if case_full_rank:
                # compute `mat_T.T @ vec_z` by `vec_p` for efficiency (page. 29)
                vec_p = mat_A.T @ vec_z
                pc1_w = pc2_w - np.hstack([vec_p - vec_z, -vec_p - vec_z])
            else:
                pc1_w = pc2_w - mat_T.T @ vec_z 

            vec_x = (pc1_w[:n] - pc1_w[n:]) / np.sqrt(2)

            # stopping criteria
            cost = np.linalg.norm(mat_A @ vec_x + mat_B @ np.abs(vec_x) - vec_c)
            if (
                cost < self.tol
                or abs(cost - prev_cost) / max(prev_cost, 1e-7) < self.rel_tol
            ):
                break

            vec_w = pc1_w
            prev_cost = cost

        if self.verbose:
            print("iterations: {}".format(i + 1))
            print("error: {:.5f}".format(cost), end="\n\n")

        return vec_x

File: MAP/solver.cpp
/**
 * Copyright 2025 Bang-Shien Chen.
 * All rights reserved. See LICENSE for the license information.
 */
#include "solver.h"

namespace MAP {

Solver::Solver(size_t max_iteration, double tolerance, double relative_tolerance, bool verbose) {
  max_iteration_ = max_iteration;
  tolerance_ = tolerance;
  relative_tolerance_ = relative_tolerance;
  verbose_ = verbose;
}

Eigen::VectorXd Solver::solve(const Eigen::MatrixXd& mat_A, const Eigen::MatrixXd& mat_B, const Eigen::VectorXd& vec_c,
                              std::optional<Eigen::VectorXd> initial) {
  int m = mat_A.rows();
  int n = mat_A.cols();

  Eigen::VectorXd vec_x;
  Eigen::VectorXd vec_w;
  Eigen::VectorXd vec_u;
  Eigen::VectorXd vec_v;
  Eigen::VectorXd vec_z;
  Eigen::VectorXd vec_p;
  Eigen::VectorXd pc1_w;
  Eigen::VectorXd pc2_w(2 * n);
  Eigen::Array<bool, Eigen::Dynamic, 1> mask;

  // compute matrix T
  Eigen::MatrixXd mat_T(m, 2 * n);
  mat_T << mat_A + mat_B, -mat_A + mat_B;
  Eigen::VectorXd cbar = std::sqrt(2.0) * vec_c;

  // check full-rank case (Lemma 2.8)
  bool case_full_rank = (m == n) && mat_B.isApprox(-Eigen::MatrixXd::Identity(n, n));

  // pre-compute Cholesky factor (page 29)
  Eigen::LLT<Eigen::MatrixXd> chol_factor(2 * (mat_A * mat_A.transpose() + mat_B * mat_B.transpose()));

  // initialization
  if (!initial.has_value()) {
    vec_w = mat_T.transpose() * chol_factor.solve(cbar);
  } else {
    if (initial->size() != 2 * n) {
      throw std::invalid_argument("Dimension of initial guess must be " + std::to_string(2 * n));
    }
    vec_w = initial.value();
  }

  double prev_cost = 1e10;
  for (size_t i = 0; i < max_iteration_; i++) {
    // Projection onto C2 (Proposition 2.2)
    Eigen::Map<Eigen::VectorXd> vec_u(vec_w.data(), n);
    Eigen::Map<Eigen::VectorXd> vec_v(vec_w.data() + n, n);
    // mask for elements where both vec_u and vec_v are negative
    Eigen::Array<bool, Eigen::Dynamic, 1> mask = (vec_u.array() < 0) && (vec_v.array() < 0);
    vec_u.array() = mask.select(0.0, (vec_u.array() >= vec_v.array()).select(vec_u.array(), 0.0));
    vec_v.array() = mask.select(0.0, (vec_v.array() >= vec_u.array()).select(vec_v.array(), 0.0));
    pc2_w << vec_u, vec_v;

    // project onto C1 (Proposition 2.1)
    vec_z = chol_factor.solve(mat_T * pc2_w - cbar);

    if (case_full_rank) {
      vec_p = mat_A.transpose() * vec_z;
      pc1_w = pc2_w - (Eigen::VectorXd(n * 2) << vec_p - vec_z, -vec_p - vec_z).finished();
    } else {
      pc1_w = pc2_w - mat_T.transpose() * vec_z;
    }

    vec_x = (pc1_w.head(n) - pc1_w.tail(n)) / std::sqrt(2.0);

    // stopping criterion
    double cost = (mat_A * vec_x + mat_B * vec_x.cwiseAbs() - vec_c).norm();
    if (cost < tolerance_ || std::abs(cost - prev_cost) / std::max(prev_cost, 1e-7) < relative_tolerance_) {
      if (verbose_) {
        std::cout << "Iterations: " << i + 1 << "\n";
        std::cout << "Error: " << cost << "\n\n";
      }
      break;
    }

    vec_w = pc1_w;
    prev_cost = cost;
  }

  return vec_x;
}

}  // namespace MAP